sfge/src/main/java/com/salesforce/rules/ops/boundary/BoundaryDetector.java

```diff
@@ -0,0 +1,67 @@
+package com.salesforce.rules.ops.boundary;
+
+import com.salesforce.exception.ProgrammingException;
+import com.salesforce.exception.UnexpectedException;
+import com.salesforce.rules.ops.visitor.LoopDetectionVisitor;
+import java.util.Optional;
+import java.util.Stack;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Tracks the beginning and end of boundaries. Each implementation of a {@link BoundaryDetector}
+ * works with a specific type of {@link Boundary}. Use {@link #peek()} to know if a boundary is
+ * currently active.
+ *
+ * @param <T> Boundary type associated with the implementation.
+ */
+public abstract class BoundaryDetector<T extends Boundary<R>, R> {
```

Comment:
Boundary detector that can be extended to other use cases. I'm imagining this can be used for Sharing Violations rule if we decide to take the path-based route.